### ISSUE TITLE
Lambda function in conditional expression not caught

### ISSUE THREAD

**Issue by thebarbershop** [PyCQA/pycodestyle#845]
## Version

```bash
$ pycodestyle --version
2.5.0
```

## Problem

The following code throws E731 as expected,

```python
if foo:
    bar = some_function
else:
    bar = lambda x:x
```

but the following doesn't.

```python
bar = some_function if foo else lambda x:x
```

In both cases, if `foo == False`, the lambda function is essentially assigned to `bar`.

[PEP8](https://www.python.org/dev/peps/pep-0008/#programming-recommendations) says

> Always use a def statement instead of an assignment statement that binds a lambda expression directly to an identifier.

Isn't this violated in the situation? Or is this a case a lambda expression isn't *directly* bound to an identifier? (because it is wrapped in a conditional expression?)

**Comment by asottile** [comment]
> Or is this a case a lambda expression isn't _directly_ bound to an identifier? (because it is wrapped in a conditional expression?)

this is my reading of this at least

There'd also be no obvious rewrite of that ternary either so I don't think it should be flagged.

(in the same way that this _technically_ assigns a lambda, but not directly)

```python
def noop(x): x

y = noop(lambda: ...)
```

**Comment by thebarbershop** [comment]
It's against the spirit of PEP8 discouraging lambda function assginments, that says

> The first form means that the name of the resulting function object is specifically 'f' instead of the generic '<lambda>'. This is more useful for tracebacks and string representations in general. The use of the assignment statement eliminates the sole benefit a lambda expression can offer over an explicit def statement (i.e. that it can be embedded inside a larger expression)

but maybe this is where we should leave the choice up to the programmer. (I'm still not very sure.)

If nobody's interested, I'm all happy to close this issue and leave it as it is.

**Comment by thebarbershop** [comment]
@asottile but wait, here's another one.

```python
def foo(bar=lambda x: x):
    ....
```

This code *directly* assigns a lambda function to a local variable (a parameter) with an exception when it is explicitly given.

Also, it seems like a very inconsistant behavior that the following code throws E731.

```python
bar = lambda x: x if foo else some_function
```

while my second example in the original post, 

```python
bar = some_function if foo else lambda x: x
```

doesn't.

**Comment by asottile** [comment]
so `bar = lambda x: x if foo else some_function` is actually equivalent to `def bar(x): return x if foo else some_function` 😉 -- `bar = (lambda x: x) if foo else some_function` does not produce any linting errors

```
$ echo 'bar = lambda x: x if foo else some_function' | astpretty /dev/stdin
Module(
    body=[
        Assign(
            lineno=1,
            col_offset=0,
            targets=[Name(lineno=1, col_offset=0, id='bar', ctx=Store())],
            value=Lambda(
                lineno=1,
                col_offset=6,
                args=arguments(
                    args=[arg(lineno=1, col_offset=13, arg='x', annotation=None)],
                    vararg=None,
                    kwonlyargs=[],
                    kw_defaults=[],
                    kwarg=None,
                    defaults=[],
                ),
                body=IfExp(
                    lineno=1,
                    col_offset=16,
                    test=Name(lineno=1, col_offset=21, id='foo', ctx=Load()),
                    body=Name(lineno=1, col_offset=16, id='x', ctx=Load()),
                    orelse=Name(lineno=1, col_offset=30, id='some_function', ctx=Load()),
                ),
            ),
        ),
    ],
)
$ echo 'bar = (lambda x: x) if foo else some_function' | astpretty /dev/stdin
Module(
    body=[
        Assign(
            lineno=1,
            col_offset=0,
            targets=[Name(lineno=1, col_offset=0, id='bar', ctx=Store())],
            value=IfExp(
                lineno=1,
                col_offset=6,
                test=Name(lineno=1, col_offset=23, id='foo', ctx=Load()),
                body=Lambda(
                    lineno=1,
                    col_offset=7,
                    args=arguments(
                        args=[arg(lineno=1, col_offset=14, arg='x', annotation=None)],
                        vararg=None,
                        kwonlyargs=[],
                        kw_defaults=[],
                        kwarg=None,
                        defaults=[],
                    ),
                    body=Name(lineno=1, col_offset=17, id='x', ctx=Load()),
                ),
                orelse=Name(lineno=1, col_offset=32, id='some_function', ctx=Load()),
            ),
        ),
    ],
)
```


I do think all of your example are _in spirit of the rule_ -- no disagreement there.  but they'd also be quite annoying as lint rules.  rewriting the parameter case requires introducing a global function, whereas replacing a simple assignment can be done without adding new names

note also that pycodestyle doesn't access an ast, it does its checks based on tokens and regexes which are much much less powerful for detecting syntax constructs

**Comment by thebarbershop** [comment]
Oops, you are right with `bar = lambda x: x if foo else some_function`. My mistake, and I also see that `bar = (lambda x: x) if foo else some_function` doesn't throw a linting error.

And I also think that it may not be a good choice to add this as a linting rule. I still think it would be better if it throws at least a warning or whatever, but it won't be easy to catch. I'm happy to leave it as it is.

Closing this issue in 24h if nobody's interested anymore, and thanks the discussion.

**Comment by asottile** [comment]
I don't mean to discourage you :)

There aren't "warnings" as far as I know, though there are "disabled by default" rules -- which maybe something could be made for this?

**Comment by thebarbershop** [comment]
Aren't there warnings? I see some like "W293 | blank line contains whitespace" or "W604 | backticks are deprecated, use ‘repr()’" in the [docs](https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes), or maybe they aren't for these type of problems.

So closing for now, because now I don't think this can be dealt by pycodestyle, but leave a comment or quote this if somebody feels the same need ¯\_(ツ)_/¯
